ibis/expr/operations/generic.py

```diff
@@ -156,25 +170,28 @@ class NotNull(UnaryOp):
         Value expression indicating whether values are not null
     """
 
-    output_type = rlz.shape_like('arg', dt.boolean)
+    output_dtype = dt.boolean
 
 
 @public
 class ZeroIfNull(UnaryOp):
-    output_type = rlz.typeof('arg')
+    output_dtype = rlz.dtype_like("arg")
 
 
 @public
 class IfNull(ValueOp):
-    """Equivalent to (but perhaps implemented differently):
+    """
+    Equivalent to (but perhaps implemented differently):
 
     case().when(expr.notnull(), expr)
           .else_(null_substitute_expr)
     """
 
     arg = rlz.any
     ifnull_expr = rlz.any
-    output_type = rlz.shape_like('args')
+
+    # we may not need this if we use this rule on valueop by default
+    output_dtype = rlz.dtype_like("args")
```

Comment:
Remove this since ValueOp's default `output_dtype` is identical with this property.

Reply:
Bump

Reply:
Updated.